crates/kumo-dmarc/src/lib.rs

```diff
@@ -0,0 +1,498 @@
+        match value {
+            Mode::Relaxed => 'r',
+            Mode::Strict => 's',
+        }
+    }
+}
+
+impl FromStr for Mode {
+    type Err = String;
+
+    fn from_str(value: &str) -> Result<Self, Self::Err> {
+        Ok(match value {
+            "r" => Self::Relaxed,
+            "s" => Self::Strict,
+            _ => return Err(format!("invalid mode {value:?}")),
+        })
+    }
+}
+
+#[derive(Clone, Copy, Debug, Eq, PartialEq)]
+struct ReportFailure {
+    all_pass: bool,
+    any_pass: bool,
+    dkim: bool,
+    spf: bool,
+}
+
+impl ToXml for ReportFailure {
+    fn serialize<W: std::fmt::Write + ?Sized>(
+        &self,
+        _: Option<instant_xml::Id<'_>>,
+        serializer: &mut instant_xml::Serializer<W>,
+    ) -> Result<(), instant_xml::Error> {
+        serializer.write_str(self)
+    }
+}
+
+impl fmt::Display for ReportFailure {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        let values = [
+            (self.all_pass, '0'),
+            (self.any_pass, '1'),
+            (self.dkim, 'd'),
+            (self.spf, 's'),
+        ];
+
+        let mut first = true;
+        for &(_, ch) in values.iter().filter(|(value, _)| *value) {
+            if !first {
+                f.write_char(':')?;
+            } else {
+                first = false;
+            }
+
+            f.write_char(ch)?;
+        }
+
+        Ok(())
+    }
+}
+
+impl FromStr for ReportFailure {
+    type Err = String;
+
+    fn from_str(value: &str) -> Result<Self, Self::Err> {
+        let mut new = Self::default();
+        for part in value.split(':') {
+            match part.trim() {
+                "0" => new.all_pass = true,
+                "1" => new.any_pass = true,
+                "d" => new.dkim = true,
+                "s" => new.spf = true,
+                _ => return Err(format!("invalid report failure {value:?}")),
+            }
+        }
+
+        Ok(new)
+    }
+}
+
+impl Default for ReportFailure {
+    fn default() -> Self {
+        Self {
+            all_pass: true,
+            any_pass: false,
+            dkim: false,
+            spf: false,
+        }
+    }
+}
+
+#[derive(Debug, Clone, Eq, PartialEq)]
+struct FeedbackAddress {
+    uri: String,
+    size: Option<u64>,
+}
+
+impl FromStr for FeedbackAddress {
+    type Err = String;
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        let s = s.trim();
+        if s.is_empty() {
+            return Err("empty feedback address {s:?}".to_owned());
+        }
+
+        let Some((uri, size)) = s.trim().rsplit_once('!') else {
+            return Ok(Self {
+                uri: s.to_owned(),
+                size: None,
+            });
+        };
+
+        let size = size.trim();
+        if size.is_empty() {
+            return Err(format!("empty size in {s:?}"));
+        }
+
+        let mut power = 0;
+        match size.chars().next_back() {
+            Some('k') => power = 10,
```

Comment:
heh, I've gone this many years without seeing this particular way of handling this sort of suffix. Neat!